File: app/src/main/java/in/shaaan/pcipharmd/FourthYear.java
package in.shaaan.pcipharmd;

import static in.shaaan.pcipharmd.AdUtil.loadNativeAd;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.ads.AdRequest;
import com.google.android.material.snackbar.Snackbar;

import java.util.HashMap;
import java.util.Map;

import in.shaaan.pcipharmd.databinding.ActivityFourthYearBinding;
import in.shaaan.pcipharmd.databinding.ContentFourthYearBinding;

public class FourthYear extends AppCompatActivity implements View.OnClickListener {
    ContentFourthYearBinding yearBinding;
    private ActivityFourthYearBinding activityFourthYearBinding;
    AdUtil adUtil = new AdUtil();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityFourthYearBinding = ActivityFourthYearBinding.inflate(getLayoutInflater());
        setContentView(activityFourthYearBinding.getRoot());
        setSupportActionBar(activityFourthYearBinding.toolbar);
        yearBinding = activityFourthYearBinding.layout4y;
        adUtil.loadInterAd(this);

        initializeUI();
        refreshAd();
    }

    private void initializeUI() {
        yearBinding.tp3.setOnClickListener(this);
        yearBinding.toxicology.setOnClickListener(this);
        yearBinding.hp.setOnClickListener(this);
        yearBinding.cp.setOnClickListener(this);
        yearBinding.biopharm.setOnClickListener(this);
        yearBinding.biostat.setOnClickListener(this);

        activityFourthYearBinding.fab.setOnClickListener(view -> Snackbar.make(view, "Like the app? Rate it on Play Store!", Snackbar.LENGTH_LONG)
                .setAction("RATE", v -> {
                    Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=in.shaaan.pcipharmd"));
                    startActivity(intent);
                }).show());

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }
    
    public void refreshAd() {
        AdRequest adRequest = new AdRequest.Builder().build();
        loadNativeAd(this, yearBinding.nativeAd41, yearBinding.nativeCard41);
        loadNativeAd(this, yearBinding.nativeAd42, yearBinding.nativeCard42);
        loadNativeAd(this, yearBinding.nativeAd43, yearBinding.nativeCard43);
    }


    @Override
    public void onClick(View view) {
        Map<Integer, String> urlMap = new HashMap<>();
        urlMap.put(R.id.tp3, "https://shaaan.github.io/pcipd/syllabus4/pt3");
        urlMap.put(R.id.toxicology, "https://shaaan.github.io/pcipd/syllabus4/toxicology");
        urlMap.put(R.id.cp, "https://shaaan.github.io/pcipd/syllabus4/cp");
        urlMap.put(R.id.hp, "https://shaaan.github.io/pcipd/syllabus4/hosp_pharm");
        urlMap.put(R.id.biopharm, "https://shaaan.github.io/pcipd/syllabus4/biopharm");
        urlMap.put(R.id.biostat, "https://shaaan.github.io/pcipd/syllabus4/biostat");

        String url = urlMap.get(view.getId());
        if (url != null) {
            adUtil.showInterAd(this, url);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_home, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            startActivity(new Intent(this, About.class));
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/in/shaaan/pcipharmd/FourthYear.kt
package `in`.shaaan.pcipharmd

import android.view.View
import androidx.appcompat.widget.Toolbar
import com.google.android.material.floatingactionbutton.FloatingActionButton
import `in`.shaaan.pcipharmd.databinding.ActivityFourthYearBinding
import `in`.shaaan.pcipharmd.databinding.ContentFourthYearBinding

/**
 * Activity for the Fourth Year syllabus.
 * What: Displays subjects for the fourth year and handles navigation to their details.
 * Why: Provides a dedicated screen for fourth-year content.
 * How: Extends `BaseYearActivity` to inherit common functionalities.
 */
class FourthYear : BaseYearActivity<ActivityFourthYearBinding, ContentFourthYearBinding>(
    ActivityFourthYearBinding::inflate
) {
    companion object {
        private const val TP3_PATH = "fourth_year/pt3"
        private const val TOXICOLOGY_PATH = "fourth_year/toxicology"
        private const val CP_PATH = "fourth_year/cp"
        private const val HP_PATH = "fourth_year/hosp_pharm"
        private const val BIOPHARM_PATH = "fourth_year/biopharm"
        private const val BIOSTAT_PATH = "fourth_year/biostat"
    }

    override fun getContentBinding(activityBinding: ActivityFourthYearBinding) =
        activityBinding.layout4y

    override fun getToolbar(activityBinding: ActivityFourthYearBinding): Toolbar =
        activityBinding.toolbar

    override fun getFab(activityBinding: ActivityFourthYearBinding): FloatingActionButton =
        activityBinding.fab

    override val subjectButtonIdToEndpointMap: Map<Int, String> = mapOf(
        R.id.tp3 to TP3_PATH,
        R.id.toxicology to TOXICOLOGY_PATH,
        R.id.cp to CP_PATH,
        R.id.hp to HP_PATH,
        R.id.biopharm to BIOPHARM_PATH,
        R.id.biostat to BIOSTAT_PATH
    )

    override fun getNativeAdPlaceholderPairs(contentBinding: ContentFourthYearBinding): List<Pair<View, View>> {
        return listOf(
            Pair(contentBinding.nativeAd41, contentBinding.nativeCard41),
            Pair(contentBinding.nativeAd42, contentBinding.nativeCard42),
            Pair(contentBinding.nativeAd43, contentBinding.nativeCard43)
        )
    }

    override fun getSubjectButtons(contentBinding: ContentFourthYearBinding): List<View> {
        return listOf(
            contentBinding.tp3, contentBinding.toxicology, contentBinding.hp,
            contentBinding.cp, contentBinding.biopharm, contentBinding.biostat
        )
    }
}
